haproxy-healthcheck/check.py

```diff
@@ -1,80 +1,93 @@
-import socket, requests
+from typing import List
 
-from _thread import *
-import threading
+import httpx
 
-print_lock = threading.Lock()
+from logger import log
 
-couchdb_servers = os.environ['COUCHDB_SERVERS'].split(',')
+couchdb_servers: List[str] = os.environ["COUCHDB_SERVERS"].split(",")
 # Example: COUCHDB_SERVERS="couchdb.1,couchdb.2,couchdb.3"
-username = os.environ['COUCHDB_USER']
-password = os.environ['COUCHDB_PASSWORD']
+username: str = os.environ["COUCHDB_USER"]
+password: str = os.environ["COUCHDB_PASSWORD"]
 
-def get_membership_endpoint(couchdb_url):
-    return f'http://{username}:{password}@{couchdb_url}:5984/_membership'
 
-def close_connection(connection):
-    print('Closing connection ...')
-    connection.close()
+def get_membership_endpoint(couchdb_url: str) -> str:
+    return f"http://{username}:{password}@{couchdb_url}:5984/_membership"
 
-def send_down_response(conn):
-    print('Sending response: Down (CouchDB cluster not okay) ...')
-    conn.send(b'down\n')
-    close_connection(conn)
 
-def send_up_response(conn):
-    print('Sending response: Up (Couchdb cluster okay) ...')
-    conn.send(b'up\n')
-    close_connection(conn)
+def _send_down_response(writer: asyncio.StreamWriter) -> None:
+    _send_and_close(writer, b"down\n")
 
-def threaded(conn):
+
+def _send_up_response(writer: asyncio.StreamWriter) -> None:
+    _send_and_close(writer, b"up\n")
+
+
+def _send_and_close(writer: asyncio.StreamWriter, message: bytes) -> None:
+    log.info("Response: %r", message)
+    writer.write(message)
+    log.info("Closing connection")
+    writer.close()
+
+
+async def is_healthy() -> bool:
+    """
+    Checks that all couchdb servers are part of the cluster
+    """
     try:
-        result = []
-        couchdb_server = couchdb_servers[0]
-        try:
-            r = requests.get(get_membership_endpoint(couchdb_server), timeout=1)
-        except Exception as e:
-            print("Exception: ", e)
-        data = r.json()
+        first_couchdb_server: str = couchdb_servers[0]
+        async with httpx.AsyncClient() as client:
+            r = await client.get(
+                get_membership_endpoint(first_couchdb_server), timeout=1
+            )
+        data: dict = r.json()
+        log.debug("Response data: %r", data)
 
-        all_nodes = sorted(data['all_nodes'])
-        cluster_nodes = sorted(data['cluster_nodes'])
+        all_nodes: List[str] = sorted(data["all_nodes"])
+        cluster_nodes: List[str] = sorted(data["cluster_nodes"])
 
         if len(all_nodes) != len(couchdb_servers):
-            print('Nodes starting up')
-            print(f'Details: all_nodes: {all_nodes}. couchdb_servers: {couchdb_servers}')
-            result.append(b'down\n')
+            log.warning("Nodes starting up")
+            log.warning(
+                f"Details: all_nodes: {all_nodes}. couchdb_servers: {couchdb_servers}"
+            )
+            return False
         elif all_nodes != cluster_nodes:
```

Comment:
Can this elif be an if instead?

Reply:
Personally I find the original control flow to have clearer intention, but ultimately it's a negligible difference and not worth going back and forth on so I'll update the PR 😃